das/daser_test.go

```diff
@@ -4,29 +4,153 @@ import (
 	"context"
 	"sync"
 	"testing"
+	"time"
+
+	"github.com/ipfs/go-datastore"
+	ds_sync "github.com/ipfs/go-datastore/sync"
+	mdutils "github.com/ipfs/go-merkledag/test"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 
 	"github.com/celestiaorg/celestia-node/service/header"
 	"github.com/celestiaorg/celestia-node/service/share"
 )
 
-func TestDASer(t *testing.T) {
-	shareServ, dah := share.RandLightServiceWithSquare(t, 16)
+// TestDASerLifecycle tests to ensure every mock block is DASed and
+// the DASer checkpoint is updated to network head.
+func TestDASerLifecycle(t *testing.T) {
+	ds := ds_sync.MutexWrap(datastore.NewMapDatastore())
+	cstore := NewCheckpointStore(ds) // we aren't storing the checkpoint so that DASer starts DASing from height 1.
 
-	randHeader := header.RandExtendedHeader(t)
-	randHeader.DataHash = dah.Hash()
-	randHeader.DAH = dah
+	mockGet, shareServ, sub := createDASerSubcomponents(t, 15, 15)
 
-	sub := &header.DummySubscriber{
-		Headers: []*header.ExtendedHeader{randHeader},
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*15)
+	t.Cleanup(cancel)
+
+	daser := NewDASer(shareServ, sub, mockGet, cstore)
+
+	err := daser.Start(ctx)
+	require.NoError(t, err)
+
+	select {
+	// wait for sample routine to finish so that it stores the
+	// latest DASed checkpoint so that it stores the latest DASed checkpoint
+	case <-daser.sampleDn:
+	case <-ctx.Done():
 	}
 
-	daser := NewDASer(shareServ, sub)
+	// load checkpoint and ensure it's at network head
+	checkpoint, err := loadCheckpoint(cstore)
+	require.NoError(t, err)
+	assert.Equal(t, int64(30), checkpoint)
+
+	err = daser.Stop(ctx)
+	require.NoError(t, err)
+}
+
+func TestDASer_catchUp(t *testing.T) {
+	ds := ds_sync.MutexWrap(datastore.NewMapDatastore())
+	cstore := NewCheckpointStore(ds)
+
+	mockGet, shareServ, _ := createDASerSubcomponents(t, 5, 0)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
+
+	daser := NewDASer(shareServ, nil, mockGet, cstore)
+
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	go func(wg *sync.WaitGroup) {
+		daser.catchUp(ctx, 2, mockGet.head) // pick random number as `from` parameter
```

Comment:
which random number? `2`?

Reply:
Yes

Reply:
![image](https://user-images.githubusercontent.com/7441766/156174277-65222eb6-f663-4e86-9f6e-d493f555da83.png)
https://3d.xkcd.com/221/

Reply:
maybe replace "random" with "arbitrary" or simply remove the comment and introduce a var `someHeight := 2` or sth